API: align default value for copy arguments with numpy at runtime (copy=False raises an error if copies cannot be avoided)

neutrinoceros opened the pull request:
### Description
This is based off #16166 and #16170 and adds a minor but crucial API change to align the default value of our `copy` arguments with whatever numpy API is installed. See #16167 for the broader context.

Alternative to #16174:
**Passing `copy=False` immediately breaks if copies cannot be avoided** (in alignment with numpy 2.0)

This is meant to be a small self-contained PR that we can ask downstream stakeholders to test from, but this will only make sense once #16166 and #16170 are merged. At that point, I'll call for feedback on the dev mailing list.

Close #16167

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

neutrinoceros commented:
Note to self (and reviewers): if this is turns out preferred over #16174, I'll need to update docstrings where due. I'm avoiding doing it now so rebasing stays as easy a possible.

pllim commented:
FWIW, I vote for this one but you should wait a bit more before pushing more changes, just in case. Thanks!

pllim commented:
Well, as long as behavior does not change for numpy 1.x, is that correct?

neutrinoceros commented:
That's right. The title is maybe misleading, but this patch just delegates all copy decisions to numpy, which, with numpy 1 installed, isn't a change at all.

taldcroft commented:
@neutrinoceros - I took some time to try getting up to speed on the huge refactors to address the numpy 2.0 `copy` issue. From what I can see this all looks good and aligns with the (good) decision to delegate the default behavior to numpy. This PR continues that effort in a few places.

My question here is process. The PR description mentions that this PR is something to "ask downstream stakeholders to test from". I would assume you mean to merge this to main and let people test with astropy dev (main). 

Overall everything I see in this PR looks good and uncontroversial at this point. However there is a conflict that needs to be resolved.

neutrinoceros commented:
> My question here is process. The PR description mentions that this PR is something to "ask downstream stakeholders to test from". I would assume you mean to merge this to main and let people test with astropy dev (main).

Actually what I meant was to offer downstream users to test the unmerged branch, because I was initially more worried about breaking stuff than I am now.

In hindsight, merging first, as you suggest, would probably be easier for everyone.

Since I haven't heard back from any one yet, I'll assume nobody's currently relying on this branch's state, so let me rebase to fix the one conflict !

mhvk commented on astropy/coordinates/representation/cylindrical.py:
```diff
@@ -154,7 +154,7 @@ class CylindricalDifferential(BaseDifferential):
     base_representation = CylindricalRepresentation
 
-    def __init__(self, d_rho, d_phi=None, d_z=None, copy=False):
+    def __init__(self, d_rho, d_phi=None, d_z=None, copy=COPY_IF_NEEDED):
```
This is actually really weird - why would the `Differential` have `copy=False` as a default when the representation does not (and indeed all representations and differentials have `copy=True`, though mostly implicitly via `BaseRep*`; but a concrete example is `PhysicsSphericalDifferential`).

It maybe should be in a different PR, but let's at least not mention this in the changelog entry...

neutrinoceros replied:
How about I revert this bit here and open a dedicated issue for it ?

mhvk replied:
Yes, good idea, let's just have a PR that changes it to `True` - this is the default for essentially all astropy classes (except `Masked`, where `False` makes much more sense).

neutrinoceros replied:
done in #16198

neutrinoceros replied:
(and reverted here)

mhvk replied:
No, it shouldn't be reverted - just use `copy=True` here!

mhvk replied:
Oops, sorry, commenting on something that's out of date! All OK.

mhvk reviewed (commented):
See long in-line comment about the changelog-entry, and a general worry that something like `Table(..., copy=None)` may not work...

It may be best, however, to worry about those other things in a different issue/PR. 6.1 is not that far away and it might be good if people's CI can start testing against this...

taldcroft commented on docs/changes/16181.other.rst:
```diff
@@ -0,0 +1,11 @@
+Public APIs that expose a ``copy`` argument and that previously set ``False``
+as a default value now use ``None`` instead if Numpy 2.0 or newer is installed.
+Default behaviour is unaffected, however explicitly passing ``copy=False`` will
+now raise an error if copying cannot be avoided.
```
> A logical question here is whether we should include how to adjust it. Should we link to astropy.utils.compat.COPY_IF_NEEDED?

I had thought the right answer is to not pass a `copy=...` argument at all, deferring to the default. But this may be upsetting to people that want to be explicit and not rely on a default.

I had also wondered about `copy=None`. The table says that the behavior is formally undefined but in practice for numpy 1.x it is the same as `False`.

neutrinoceros replied:
I'll adjust the note, and I agree it belongs in the what's new.

> A logical question here is whether we should include how to adjust it. Should we link to `astropy.utils.compat.COPY_IF_NEEDED`?

As far as I understand, this module isn't considered public API (we just removed `PYTHON_LT_*` constants from it without a warning), so I wouldn't want to encourage downstream to rely on it. That said, we could recommend copying its implementation (the tricky part, I think, is just in how to properly check numpy's version at runtime using `packaging.version`). What do you think ?

> A more pragmatic question is whether we need tests to check that things do not percolate badly.

Could you clarify what you mean here ? From my standpoint, the "percolating badly" scenario would be #16174 😅 .

neutrinoceros replied:
Oops, going on a race condition with @taldcroft .

> I had thought the right answer is to not pass a `copy=...` argument at all, deferring to the default. But this may be upsetting to people that want to be explicit and not rely on a default.

Well it *is* documented as mostly for internal usage, and people who want to be explicit about *not* copying have no recourse to do so in numpy 1, so I'm not too worried.

> I had also wondered about `copy=None`. The table says that the behavior is formally undefined but in practice for numpy 1.x it is the same as False.

_unless you're using a type checker_, but essentially yes ! We could even recommend using a plain `None` as a backward compatible and cheap replacement.

mhvk replied:
I don't think one can just use `copy=None` on all numpy versions:
```
In [2]: np.__version__
Out[2]: '1.24.2'

In [3]: np.array([1, 2, 3], copy=None)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In [3], line 1
----> 1 np.array([1, 2, 3], copy=None)

ValueError: NoneType copy mode not allowed.
```

This is also what I meant by tests. Since we use `copy=False` basically only internally, I'm not so sure that we cover most cases users might do. I *think* we should be OK, though, since in most cases one would not be asking for a dtype change, so, say, `SkyCoord(ra_array, dec_array, copy=False)` should just continue to work.

Anyway, bottom line to me seems that we should just try to make sure this is out when numpy 2.0 comes out, so that we can get reports from users. Indeed, perhaps this is the one time that we should ask people to run their test suites on a point release...

neutrinoceros replied:
> I don't think one can just use copy=None on all numpy versions:

Oh gosh, how did I not check my assumptions... I'll edit the PR description, thanks !

> perhaps this is the one time that we should ask people to run their test suites on a point release...

just to be clear, you mean astropy 6.1.0 here (not 6.0.1), right ?

neutrinoceros replied:
I included Marten's suggestion to rephrase the changlog, and I've removed ``Table`` from the list. I actually don't know why I included it at first, probably a mistake !
I'll work on the what's new in the morning so hopefully this will be ready to review again tomorrow !

mhvk replied:
> just to be clear, you mean astropy 6.1.0 here (not 6.0.1), right ?

Depends on whether we'll still do a 6.0.x with numpy 2.0 support. Haven't we been backporting most things for that? (At least, things like new ufuncs, etc.) I think we need to have *some* public astropy version that works with numpy 2.0 when that comes out...

neutrinoceros replied:
Yes, 6.0.x should be compatible already and hopefully won't require more changes, but we haven't backported this PR's family of patches, which is why I got confused. Anyway, the window between 2.0.0rc1 and final should be long enough for us to land a compatible release.

mhvk replied:
Ah, yes, I guess we should backport those PRs that ensure `6.0.x` passes on numpy 2.0...

neutrinoceros replied:
It'll probably be easier to do in bulk. I suggest we move this discussion to #16167

neutrinoceros replied:
Actually I think there are conflicting desires here: if we're adding a what's new, then backporting (along with supporting PRs) becomes impossible (patch releases don't have what's news).
So, the easiest way to resolve this tension would probably be to not have a what's new, and instead make the changelog entry even longer. As long as we make sure that the release team is aware of the importance of this patch and can properly emphasize it as part of our numpy 2.0 compatibility layer, it should be good ?

mhvk replied:
Yes, makes sense!

taldcroft replied:
I really think we should only release all these changes in 6.1.0 and have a What's New.

If there is a chance that numpy 2.0 will get released before early May then we should move the astropy release forward. Having these changes in a bug fix release is (IMO) totally incompatible with our policy for such releases. Despite a lot of effort and review, I think it is credible that the numpy 2.0 compatibility changes will break some existing code somewhere. Users don't expect this for a bug fix release.

mhvk replied:
OK, that makes sense too. So, we basically will make clear that astropy 6.0 requires numpy < 2.

I had vaguely considered the idea of an early 6.1 as well - seems worth a new issue! #16200

neutrinoceros replied:
ok then I'll make a what's new from the changelog

neutrinoceros commented:
rebased again to fix an implicit conflict with https://github.com/astropy/astropy/pull/16198

mhvk commented on docs/changes/16181.other.rst:
```diff
@@ -0,0 +1,30 @@
+instances of ``False`` be replaced with a ``COPY_IF_NEEDED`` constant defined
+as follow
+
+    def NUMPY_GE_2_0() -> bool:
```
Can we just recommend a simpler version?
```
COPY_IF_NEEDED = False if np.__version__[0] == "1" else None
```

neutrinoceros replied:
Given how these compatibility layer tend to be left to rot in unmaintained software, I'm slightly uneasy to recommend something that would break in the event of numpy going to 3.0. I'm also hesitant to recommend more than one way to do it, which is why I went with the more robust alternative.

neutrinoceros replied:
hum, on second thought I think this would only break if there was a Numpy 10.0 ... and by then, anything that still has compatibility code for Numpy 1.x is likely to be broken anyway so, let's just roll with it !

neutrinoceros replied:
and on **third** thought, we can make it even more portable at no real addtional cost:
```python
COPY_IF_NEEDED = False if np.__version__.startswith("1.") else None
```

mhvk reviewed (approved):
I'm happy with this! I have a suggestion to simplify the version check but OK to leave as is too.

neutrinoceros commented:
I moved the large changelog fragment to a What's New section, and replaced the changelog with a one liner version of its past self. Should be good now ?

neutrinoceros commented:
Failure in devdeps seem unrelated but it's still kinda scary (and unfamiliar to me).

mhvk commented on docs/changes/16181.other.rst:
```diff
@@ -0,0 +1,3 @@
+Change the default value of ``copy`` arguments in public APIs from ``False`` to
+``None`` if Numpy 2.0 or newer is installed.
+For details, see :ref:`What's New in Astropy 6.1?<_whatsnew-6.1-copy-semantics>`_
```
I think we generally try to avoid putting links in changelog fragments (because they are more likely to get out of date), but I'm not sure whether this one is OK - let's ping @pllim!

pllim replied:
Yeah:

1. That ref is wrong. Should have used `whatsnew-6.1-copy-semantics`.
2. I also remember some past pain where we linked to something in What's New but then the link broke after release. But I don't remember exactly what broke. @saimn , do you?

pllim replied:
p.s. Maybe it is not the change log fragment but the what's new. See https://github.com/astropy/astropy/pull/16181#issuecomment-2000243710

pllim replied:
Ah, that would do it.

neutrinoceros replied:
thank you ! should be fixed now !

mhvk reviewed (approved):
Looks great, modulo the one worry about the link in the changelog fragment. Approving assuming that is actually OK.

pllim commented:
Re: devdeps -- Looks like transient network error but I restarted it, just in case. I also added `Extra CI` to play safe.

mhvk commented:
devdeps failure is yet another bunch of str ufuncs being added to numpy 2.0 (probably the last bunch to come...)

doc link test I didn't quite have the patience to try to understand what exactly went wrong - rather a lot of less relevant information to parse through...

pllim commented:
linkcheck isn't related, so you can ignore that one.

RTD warning is related:

```whatsnew/index
WARNING: :347: (WARNING/2) Mismatch: both interpreted text role prefix and reference suffix.
```

neutrinoceros commented:
@mhvk thanks ! I fixed the devdeps failure in https://github.com/astropy/astropy/pull/16205

neutrinoceros commented:
rebased to fix a merge conflict

mhvk commented:
failures due to #16251

pllim commented on docs/changes/16181.other.rst:
```diff
@@ -0,0 +1,3 @@
+Change the default value of ``copy`` arguments in public APIs from ``False`` to
+``None`` if Numpy 2.0 or newer is installed.
+For details, see :ref:`What's New in Astropy 6.1?<_whatsnew-6.1-copy-semantics>`
```
This failed RTD build with a warning:

```
WARNING: undefined label: '_whatsnew-6.1-copy-semantics'
```

pllim replied:
Also I don't think intersphinx is going to last forever in What's New as we "archive" the page at release time. For example, please see https://github.com/astropy/astropy/blob/main/docs/whatsnew/6.0.rst

neutrinoceros replied:
Thanks for pointing it out, how can I fix this ? Should I just refer to the what's new page in text and avoid any linking ?

pllim replied:
I am okay with a mention without any link. What do you think, @mhvk ?

mhvk replied:
Agreed, let's not over{th|l}ink.

pllim commented on docs/changes/16181.other.rst:
```diff
@@ -0,0 +1,3 @@
+Change the default value of ``copy`` arguments in public APIs from ``False`` to
+``None`` if Numpy 2.0 or newer is installed.
+For details, see the What's New page for Astropy 6.1 .
```
```suggestion
For details, see the "Copy semantics" section on the What's New page for Astropy 6.1 .
```

neutrinoceros commented:
I really can't understand how docs are still failing (I can't even find where the error is in the log)
Other failures are unrelated and already addressed.

astrofrog commented:
I can try and investigate this morning

astrofrog commented:
The RTD docs build is fine, the one that is failing is the link check, but I think these are unrelated transient issues. The log is hard to parse but by grepping ``' broken '`` I find:

```
2024-04-03T22:15:47.3818420Z (io/fits/usage/headers: line  336) broken    https://heasarc.gsfc.nasa.gov/docs/heasarc/ofwg/docs/ofwg_recomm/r13.html - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2024-04-03T22:15:47.3895910Z (api/astropy.units.format.OGIP: line    2) broken    https://heasarc.gsfc.nasa.gov/docs/heasarc/ofwg/docs/general/ogip_93_001/ - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2024-04-03T22:15:47.4963577Z (io/fits/appendix/faq: line  841) broken    https://heasarc.gsfc.nasa.gov/docs/software/fitsio/c/c_user/node97.html - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2024-04-03T22:15:47.5889324Z (io/fits/appendix/faq: line  139) broken    https://heasarc.gsfc.nasa.gov/docs/software/ftools/fitsverify/ - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2024-04-03T22:15:47.6081347Z (io/fits/appendix/faq: line  583) broken    https://heasarc.gsfc.nasa.gov/fitsio/fitsio.html - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
2024-04-03T22:15:57.2591337Z (   timeseries/io: line    9) broken    https://tess.gsfc.nasa.gov/ - ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

All are ``.nasa.gov``-related. We should perhaps (in a separate PR) enable quiet mode for the link check as suggested in https://github.com/sphinx-doc/sphinx/issues/8791 as this will then only show the broken links, not all the valid ones.

astrofrog reviewed (approved):
Failures are unrelated, looks good to me!
